ngraph/core/src/pass/make_stateful.cpp

```diff
@@ -0,0 +1,97 @@
+#include <openvino/op/util/variable.hpp>
+#include <openvino/opsets/opset8.hpp>
+
+using namespace std;
+using namespace ngraph;
+using namespace opset8;
+using namespace op::util;
+
+namespace {
+string generate_variable_name(const shared_ptr<Parameter>& param, const shared_ptr<Result>& res, uint64_t idx) {
+    return param->get_friendly_name() + res->get_friendly_name() + std::to_string(idx);
+}
+}  // namespace
+
+bool ov::pass::MakeStateful::run_on_function(std::shared_ptr<ngraph::Function> f) {
+    VariableVector variables;
+    SinkVector sinks;
+    uint64_t idx = 0;
+    for (const auto& pair : m_pairs_to_replace) {
+        const auto& param = pair.first;
+        const auto& res = pair.second;
+
+        NGRAPH_CHECK(param->get_partial_shape().is_static(),
+                     "Shape of Parameter ",
+                     param->get_friendly_name(),
+                     " must be static. MakeStateful transformation doesn't support dynamic shapes.");
+
+        const auto& target_inputs = param->get_output_target_inputs(0);
+
+        // Create Variable
+        std::string var_name = generate_variable_name(param, res, idx++);
+        auto variable = std::make_shared<Variable>(VariableInfo{PartialShape::dynamic(), element::dynamic, var_name});
+        variables.push_back(variable);
+
+        // Create ReadValue
+        auto const_zero = make_shared<Constant>(param->get_element_type(), param->get_shape(), 0);
+        auto read_val = make_shared<ReadValue>(const_zero, variable);
+        for (const auto& target_in : target_inputs) {
+            target_in.replace_source_output(read_val->output(0));
+        }
+        copy_runtime_info(param, read_val);
+
+        // Create Assign
+        const auto& input_to_res = res->input_value(0);
+        input_to_res.remove_target_input(res->input(0));
+        auto assign = make_shared<Assign>(input_to_res, variable);
+        copy_runtime_info(res, assign);
+
+        // Update Function
+        sinks.push_back(assign);
+        f->remove_result(res);
+        f->remove_parameter(param);
+        assign->add_control_dependency(read_val);
+    }
+    f->add_variables(variables);
+    f->add_sinks(sinks);
+    return true;
+}
+
+ov::pass::MakeStateful::ParamResPairs ov::pass::MakeStateful::find_param_results_by_names(
+    const shared_ptr<ngraph::Function>& func,
+    const std::map<std::string, std::string>& param_res_names) {
+    ParamResPairs pairs_to_replace;
+    const auto& params = func->get_parameters();
+    const auto& results = func->get_results();
+
+    // find corresponding param and result by name and add to the list
+    for (const auto& param_res : param_res_names) {
+        const auto& param_name = param_res.first;
+        const auto& res_name = param_res.second;
+
+        auto param = std::find_if(params.begin(), params.end(), [&](const std::shared_ptr<ngraph::Node>& node) {
+            return node->get_friendly_name().find(param_name) != std::string::npos;
```

Comment:
Why are we searching like this? but not just `node->get_friendly_name() == param_name`?

Reply:
std::string find will also return true for sub-string match. I think this is not expected.

Reply:
done